File: src/openrgb_flowers/effects/effect_engine_factory.py
"""Factory for instantiating lighting effect engines following SOLID Open/Closed principle."""
from __future__ import annotations
from typing import Optional

from openrgb_flowers.core.interfaces.i_effect_engine import IEffectEngine
from openrgb_flowers.core.interfaces.i_layout_provider import ILayoutProvider
from openrgb_flowers.core.interfaces.i_color_palette import IColorPalette
from openrgb_flowers.core.models.effect_config import EffectConfig
from openrgb_flowers.effects.blooming_engine import BloomingEngine
from openrgb_flowers.effects.random_blend_engine import RandomBlendEngine
from openrgb_flowers.effects.palettes.palette_registry import PaletteRegistry


class EffectEngineFactory:
    """Factory responsible for instantiating effect engines."""

    EFFECT_BLOOMING = "blooming"
    EFFECT_RANDOM_BLEND = "random_blend"

    AVAILABLE_EFFECTS = [
        EFFECT_BLOOMING,
        EFFECT_RANDOM_BLEND,
    ]

    @classmethod
    def create_engine(
        cls,
        effect_name: str,
        config: EffectConfig,
        layout_provider: ILayoutProvider,
        palette: Optional[IColorPalette] = None,
    ) -> IEffectEngine:
        """Instantiates the requested effect engine.

        Args:
            effect_name: Identifier ('blooming', 'random_blend', etc.)
            config: Effect configuration instance
            layout_provider: Target hardware layout provider
            palette: Optional custom palette instance (defaults to config.palette_name)
        """
        clean_name = effect_name.strip().lower().replace("-", "_").replace(" ", "_")

        resolved_palette = palette or PaletteRegistry.get(config.palette_name)

        if clean_name in ("random_blend", "random", "blend", "chaos", "matrix_blend", "all_keys"):
            return RandomBlendEngine(
                config=config,
                layout_provider=layout_provider,
                palette=resolved_palette,
            )

        # Default fallback is organic blooming
        return BloomingEngine(
            config=config,
            layout_provider=layout_provider,
            palette=resolved_palette,
        )

File: src/effects/effect_engine_factory.cpp
#include "openrgb_flowers/effects/effect_engine_factory.hpp"
#include "openrgb_flowers/effects/blooming_engine.hpp"
#include "openrgb_flowers/effects/random_blend_engine.hpp"

#include <algorithm>
#include <cctype>

namespace openrgb_flowers::effects {

namespace {

std::string to_lower(std::string s) {
    std::transform(s.begin(), s.end(), s.begin(), [](unsigned char c) {
        return static_cast<char>(std::tolower(c));
    });
    return s;
}

} // anonymous namespace

std::unique_ptr<core::interfaces::IEffectEngine> EffectEngineFactory::create_engine(
    const std::string& effect_name,
    core::models::EffectConfig config,
    std::shared_ptr<core::interfaces::ILayoutProvider> layout,
    std::shared_ptr<core::interfaces::IColorPalette> palette,
    std::shared_ptr<core::interfaces::IBlendStrategy> blend
) {
    std::string key = to_lower(effect_name);
    if (key == "random_blend" || key == "random" || key == "blend" || key == "mosaic") {
        return std::make_unique<RandomBlendEngine>(std::move(config), std::move(layout), std::move(palette));
    }
    return std::make_unique<BloomingEngine>(
        std::move(config), std::move(layout), std::move(palette), std::move(blend)
    );
}

} // namespace openrgb_flowers::effects
